doc/design/parallel-node-config.md

```diff
@@ -0,0 +1,195 @@
+
+### API Extensions
+
+#### Option 1: extend existing CR SriovNetworkPoolConfig
+SriovNetworkPoolConfig is used only for OpenShift to provide configuration for
+OVS Hardware Offloading. We can extend it to add configuration for the drain
+pool. E.g.:
+
+```yaml
+apiVersion: v1
+kind: SriovNetworkPoolConfig
+metadata:
+name: pool-1
+namespace: network-operator
+spec:
+priority:  44
+nodeSelectorTerms:
+- matchExpressions:
+    - key: some-label
+      operator: In
+      values:
+        - val-2
+- matchExpressions:
+    - key: other-label
+      operator: "Exists"
+      DrainConfig:
+      maxUnavailable: 5
+      ovsHardwareOffloadConfig: {}
+```
+
+Default pool should be always created with the lowest priority:
+```yaml
+apiVersion: v1
+kind: SriovNetworkPoolConfig
+metadata:
+name: default
+namespace: network-operator
+spec:
+priority:  100
+DrainConfig:
+maxUnavailable: 1
+```
+
+#### Option 2: extend SriovOperatorConfig CRD
+We can extend SriovOperatorConfig CRD to include drain pools configuration. E.g.:
+
+```yaml
+apiVersion: sriovnetwork.openshift.io/v1
+kind: SriovOperatorConfig
+metadata:
+name: default
+namespace: network-operator
+spec:
+# Add fields here
+enableInjector: false
+enableOperatorWebhook: false
+configDaemonNodeSelector: {}
+disableDrain: false
+drainConfig:
+- name: default
+  maxUnavailable: 1
+  priority: 100 # the lowest priority
+- name: pool-1
+  maxUnavailable: 5
+  priority: 44
+  # empty nodeSelectorTerms means 'all nodes'
+  nodeSelectorTerms:
+  - matchExpressions:
+    - key: some-label
+      operator: In
+      values:
+      - val-1
+      - val-2
+  - matchExpressions:
+    - key: other-label
+      operator: "Exists"
+```
+
+#### Option 3:  New CRD
+Add new `DrainConfiguration`CRD with fields mentioned in previous options.
+We can extend SriovOperatorConfig CRD to include drain pools configuration. E.g.:
+```yaml
+apiVersion: sriovnetwork.openshift.io/v1
+kind: SriovDrainConfig
+metadata:
+name: default
+namespace: network-operator
+spec:
+maxUnavailable: 1
+priority: 100 # the lowest priority
+# empty nodeSelectorTerms means 'all nodes'
+nodeSelectorTerms:
+- matchExpressions:
+- key: some-label
+operator: In
+``` 
+
+### Implementation Details/Notes/Constraints
+The implementation of this feature is very complex and requires a lot of changes to different parts of SR-IOV Network
+Operator. To not introduce breaking changes we have to split this effort to several phases:
+* implement new Drain controller without API changes:
+  it will proceed only one node configuration at the same time and doesn't require API changes
+* introduce new API changes to support pool of nodes to be drained in a parallel:
+  at this phase we introduce new API from one of the proposed options above and modify Drain controller to watch for
+  the specified CRs and proceed drain in a parallel per node pool configuration
+* drop `NodeDrainAnnotation` usage and move this annotation values into the `SriovNetworkNodeState` object
```

Comment:
why do we want to do this last ? and not part of the initial change ? seems like un-needed work to first implement using node annot and then move them to a different k8s obj

Reply:
agree we should move this as part of the first change.

we much check how this will work in case the operator upgrade is done in the middle of the configuration.